jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -708,6 +710,133 @@ ecma_concat_ecma_strings (ecma_string_t *string1_p, /**< first ecma-string */
   return string_desc_p;
 } /* ecma_concat_ecma_strings */
 
+#ifdef JERRY_ENABLE_ERROR_MESSAGES
+
+/**
+ * Format ecma-values.
+ *
+ * @return a newly created ecma-string
+ */
+ecma_string_t *
+ecma_format_ecma_values (const char *format, /**< format string */
+                         ...) /**< ecma-values */
+{
+  JERRY_ASSERT (format != NULL);
+
+  va_list args;
+
+  va_start (args, format);
+
+  ecma_string_t *string_p = ecma_get_magic_string (LIT_MAGIC_STRING__EMPTY);
+  ecma_string_t *string1_p;
+  ecma_string_t *string2_p;
+
+  const char *start_p = format;
+  const char *end_p = format;
+
+  while (*end_p)
+  {
+    if (*end_p == '%')
+    {
+      /* Concat template string. */
+      if (end_p > start_p)
+      {
+        string1_p = string_p;
+        string2_p = ecma_new_ecma_string_from_utf8 ((const lit_utf8_byte_t *) start_p,
+            (lit_utf8_size_t) (end_p - start_p));
+        string_p = ecma_concat_ecma_strings (string1_p, string2_p);
+        ecma_deref_ecma_string (string1_p);
+        ecma_deref_ecma_string (string2_p);
+      }
+
+      /* Convert an argument to string without side effect. */
+      ecma_string_t *arg_string_p;
+      const ecma_value_t arg_val = va_arg (args, ecma_value_t);
+      if (unlikely (ecma_is_value_object (arg_val)))
+      {
+        ecma_object_t *arg_object_p = ecma_get_object_from_value (arg_val);
+        lit_magic_string_id_t class_name = ecma_object_get_class_name (arg_object_p);
+        arg_string_p = ecma_get_magic_string (class_name);
+      }
+      else
+      {
+        if (ecma_is_value_string (arg_val))
+        {
+          arg_string_p = ecma_get_string_from_value (arg_val);
+          ecma_ref_ecma_string (arg_string_p);
```

Comment:
Because `ecma_get_magic_string` will increase the reference count, while `ecma_get_string_from_value` won't. To make algorithm consistent, we increase reference count for the later case, and dereference them after concatation.